### PR TITLE
fix(pymongo): support 3.12.0

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
## Description

With pymongo 3.12 an API that was used for hooking span creation [was
renamed](https://github.com/mongodb/mongo-python-driver/compare/3.11.4...3.12.0#diff-d5e3bba0b330986914b87fdbdbb9fe506eaf7f0d0cdd8b2f92081ff23620deaaL70). (Specifically this commit: https://github.com/mongodb/mongo-python-driver/commit/c8f32a7a376e027cbe30dcb8df9bf05778e61248)

The code was also cleaned up a bit by using the span context manager
rather than a try..finally which also means that exception information
will be included on spans.

## Checklist
- [x] Added to the correct milestone.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
  - existing tests should be sufficient but the test matrix was updated to include the new version of pymongo
- ~[x] Library documentation is updated.~
- ~[x] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).~

### CODE REVIEW

**Comment by Kyle-Verhoog** [comment]
@mergifyio backport 0.50

File: ddtrace/contrib/pymongo/client.py [hunk 1]
```diff
@@ -114,35 +114,29 @@ def _datadog_trace_operation(self, operation):
...
+    def run_operation(self, sock_info, operation, *args, **kwargs):
+        with self._datadog_trace_operation(operation) as span:
+            result = self.__wrapped__.run_operation(sock_info, operation, *args, **kwargs)
+            if result and result.address:
+                set_address_tags(span, result.address)
+            return result
 
-        try:
+    # Pymongo >= 3.9, <3.12
+    def run_operation_with_response(self, sock_info, operation, *args, **kwargs):
```
Comment (brettlangdon) [inline ddtrace/contrib/pymongo/client.py:126]: should we have these within if statements?

what happens if we add `run_operation_with_response` to version >=3.12?

Comment (Kyle-Verhoog) [reply ddtrace/contrib/pymongo/client.py:126]: We're just adding functions that don't get used.

I suppose the worst that could happen is a user is using this API and doesn't upgrade their code. In that case they'll see the wrapper function in the stack trace as `result = self.__wrapped__.run_operation_with_response(sock_info, operation, *args, **kwargs)` will raise.

I'll see if it's easy enough to version it though :+1:

Comment (Yun-Kim) [reply ddtrace/contrib/pymongo/client.py:126]: Just to clarify, `run_operation` and `run_operation_with_response` are the same API, just renamed for different versions of Pymongo (>=3.12 and >=3.9 respectively), right? I.e. `run_operation_with_response` won't be called for versions >=3.12?

File: ddtrace/contrib/pymongo/client.py [hunk 2]
```diff
@@ -114,35 +117,30 @@ def _datadog_trace_operation(self, operation):
...
-            result = self.__wrapped__.run_operation_with_response(sock_info, operation, *args, **kwargs)
-
-            if result and result.address:
-                set_address_tags(span, result.address)
-            return result
-        finally:
-            span.finish()
-
+    # Pymongo >= 3.12
+    if VERSION >= (3, 12, 0):
```
Comment (brettlangdon) [inline ddtrace/contrib/pymongo/client.py:121]: thanks, makes me feel better

**Review by brettlangdon — DISMISSED** [review]

File: ddtrace/contrib/pymongo/client.py (hunk 1 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/pymongo/client.py]: Correct: https://github.com/mongodb/mongo-python-driver/commit/c8f32a7a376e027cbe30dcb8df9bf05778e61248

**Review by Yun-Kim — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

File: ddtrace/contrib/pymongo/client.py (hunk 1 above)
Comment (stradigi-David-Malvar) [reply ddtrace/contrib/pymongo/client.py]: Question: How can we use `with self._datadog_trace_operation(operation) as span:`  when the function isn't a context manager?

Comment (Kyle-Verhoog) [reply ddtrace/contrib/pymongo/client.py]: @stradigi-David-Malvar it's the return value of the function that has to be a context manager. `_datadog_trace_operation` returns a span which is a context manager.

**Pull request merged**